File: old/source/frontend/logger.cpp
#include <frontend/logger.hpp>
#include <standard/io.hpp>

// Error message
void Logger::error(string message)
{
    had_error = true;
    write(ConsoleStream::Error, "[", path, "] Error: ", message, "\n");
}

// Error at token
void Logger::error(Token &token, string message)
{
    had_error = true;

    write(ConsoleStream::Error, "[", path, ":", token.line, ":", token.column, "] Error");

    if (token.kind == TokenKind::End)
        write(ConsoleStream::Error, " at end");
    else if (token.kind == TokenKind::Error)
        write(ConsoleStream::Error, ": ");
    else
        write(ConsoleStream::Error, " at '", token.to_string(), "': ");

    writeln(ConsoleStream::Error, message);
}

File: source/Suteki.Compiler/Logger.cs
namespace Suteki.Compiler;

using System;

public class Logger
{
    public string Path;

    // Show error
    public void Error(string message)
    {
        Config.HadError = true;
        
        if (!Config.CanLog)
            return;

        Console.Error.WriteLine($"[{Path}] Error: {message}");
    }

    // Show error at token
    public void Error(Token token, string message)
    {
        Config.HadError = true;

        if (!Config.CanLog)
            return;

        Console.Error.Write($"[{Path}:{token.Line}:{token.Column}] Error");

        switch (token.Kind)
        {
            case TokenKind.End:
            {
                Console.Error.Write(" at end: ");
                break;
            }

            case TokenKind.Error:
            {
                Console.Error.Write(": ");
                break;
            }

            default:
            {
                Console.Error.Write($" at '{token.Content}': ");
                break;
            }
        }

        Console.Error.WriteLine(message);
    }
}